File: 2022/day12.clj
(ns day12
  (:require [clojure.string :as str]))

(defn parse [f]
  (->> f slurp str/split-lines (mapv #(mapv identity %))))

(defn inbounds? [grid [x y]]
  (let [xmax (-> grid first count dec)
        ymax (-> grid count dec)]
    (and (<= 0 x xmax) (<= 0 y ymax))))

(defn lookup [grid [x y]]
  (-> grid (get y) (get x)))

(defn lookupn [grid coord]
  (let [c (lookup grid coord)]
    (int ({\S \a \E \z} c c))))

(defn nextmoves [grid [x y :as coord]]
  (let [cur      (lookupn grid coord)
        possible [[x (inc y)] [x (dec y)] [(inc x) y] [(dec x) y]]]
    (->> possible (filter (partial inbounds? grid)) (filter #(<= (- (lookupn grid %) cur) 1)))))

(defn findall [grid char]
  (for [y     (-> grid count range)
        x     (-> grid first count range)
        :when (= char (lookup grid [x y]))]
    [x y]))

(defn minpath [grid start end]
  (loop [q    (conj clojure.lang.PersistentQueue/EMPTY [start, 0])
         seen #{start}]
    (let [[point n] (peek q)
          q'        (pop q)
          moves     (some->> point (nextmoves grid) (remove seen))
          seen'     (into seen moves)
          q'        (into q' (map (fn [m] [m (inc n)]) moves))]
      (cond
        (empty? q)    nil
        (= point end) n
        :else         (recur q' seen')))))

(defn solve1 [f]
  (let [grid  (parse f)
        start (first (findall grid \S))
        end   (first (findall grid \E))]
    (minpath grid start end)))

(defn solve2 [f]
  (let [grid   (parse f)
        starts (concat (findall grid \a) (findall grid \S))
        end    (first (findall grid \E))]
    (reduce min (keep #(minpath grid % end) starts))))

(println (solve1 "input12ex1.txt"))
(println (solve1 "input12.txt"))
(println (solve2 "input12ex1.txt"))
(println (solve2 "input12.txt"))

(comment
  (solve1 "input12ex1.txt")
  (solve1 "input12.txt")
  (solve2 "input12ex1.txt")
  (solve2 "input12.txt"))

File: 2022/day12.ml
open Base
open Stdio

module IntPair = struct
  module T = struct
    type t = int * int [@@deriving sexp_of, compare, hash, equal]
  end
  include T
  include Base.Comparator.Make(T)
end

let parse s =
  let lines = s |> String.split_lines |> Array.of_list in
  let grid = Array.map ~f:String.to_array lines in
  grid

let find_points (grid : char array array) (char : char) =
  let res = ref [] in
  for y = 0 to Array.length grid - 1 do
    for x = 0 to Array.length grid.(y) - 1 do
      if Char.(grid.(y).(x) = char)
      then res := (x, y) :: !res
    done
  done;
  !res

let find_point grid char = find_points grid char |> List.hd_exn

let minpath (grid : char array array) start goal =
  let xmax = Array.length grid.(0) in
  let ymax = Array.length grid in
  let elevation (x, y) = Char.to_int (match grid.(y).(x) with | 'S' -> 'a' | 'E' -> 'z' | c -> c) in
  let inbounds = fun (x, y) -> x >= 0 && x < xmax && y >= 0 && y < ymax in
  let q = Queue.create () in
  Queue.enqueue q (start, 0);
  let seen = Set.of_list (module IntPair) [start] in
  let rec bfs = fun seen ->
    match Queue.dequeue q with
    | None -> None
    | Some (point, n) when IntPair.equal point goal -> Some n
    | Some ((x, y) as point, n) ->
    let moves =
      [(x + 1, y); (x - 1, y); (x, y + 1); (x, y - 1)]
      |> List.filter
         ~f:(fun move ->
              not (Set.mem seen move) &&
              inbounds move &&
              elevation move - elevation point <= 1)
    in
    let seen' = Set.union seen (Set.of_list (module IntPair) moves) in
    Queue.enqueue_all q (List.map ~f:(fun m -> (m, n + 1)) moves);
    bfs seen' in
  bfs seen

let minpath1 grid =
  let start = find_point grid 'S' in
  let goal  = find_point grid 'E' in
  minpath grid start goal |> Option.value_exn

let minpath2 grid =
  let starts = (find_points grid 'S') @ (find_points grid 'a') in
  let goal = find_point grid 'E' in
  starts |> List.filter_map ~f:(fun s -> minpath grid s goal)
         |> List.min_elt ~compare:Int.compare
         |> Option.value_exn

let inp = "input12.txt" |> In_channel.read_all |> parse
(* let () = inp |> [%show: char array array] |> print_endline*)
let () = inp |> minpath1 |> printf "%i\n"
let () = inp |> minpath2 |> printf "%i\n"
